### PR TITLE
API work package description/activity comments formatting

### PR DESCRIPTION
**Pull request by myabc**
https://www.openproject.org/work_packages/13560

### CODE REVIEW

**Comment by myabc** [comment]
@ulferts  @hschink would you mind having a look at this? since you know the codebase better than I do, you might be aware of any potential complications.

File: lib/open_project/text_formatting.rb [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+      end
+      return '' if text.blank?
+
+      edit = !!options.delete(:edit)
+      # don't return html in edit mode when textile or text formatting is enabled
+      return text if edit
+      project = options[:project] || @project || (obj && obj.respond_to?(:project) ? obj.project : nil)
+      only_path = options.delete(:only_path) == false ? false : true
+
+      text = Redmine::WikiFormatting.to_html(Setting.text_formatting, text, :object => obj, :attribute => attr, :edit => edit) { |macro, args| exec_macro(macro, obj, args, :view => self, :edit => edit) }
```
Comment (ulferts) [inline lib/open_project/text_formatting.rb:82]: This one will break as `exec_macro` will call render which is only defined if the module is included into a view. 

<pre>
macro (undefined method `render' for #<API::V3::WorkPackages::WorkPackageModel:0x00000107854c30>)
</pre>

Comment (ulferts) [inline lib/open_project/text_formatting.rb:82]: Just found out that this only applies to the `timelines` macro.

**Comment by ulferts** [comment]
There is one general concern I have that should not prevent us from merging this one as I like it very much but that should make us careful when doing it. I am not sure of whether some plugins patch the rendering. Nothing comes to my mind when thinking about it but either we run it through our CI with a couple of plugins or we prepare to fix it once merged and should then not merge it too late.

File: lib/open_project/text_formatting.rb (hunk 1 above)
Comment (ulferts) [inline lib/open_project/text_formatting.rb:82]: @myabc and the rest of the gang agreed to show some message along the lines of "not supported in this context" if we render outside a view context.

**Comment by ulferts** [comment]
Just noticed that this would also have to be applied to the comments (activity) section.

**Comment by myabc** [comment]
@ulferts 

> Just noticed that this would also have to be applied to the comments (activity) section.

wasn't this done in da82c963?

**Comment by ulferts** [comment]
@myabc: Only for the attributes that where changes. But the comment (notes in AR) where not textilized there as this would require what you did in this PR.

We will have to repeat what you did in the work_package_model/representer for the activity_model/representer.

**Comment by myabc** [comment]
Note: I've made HTML the default for both description and comments (and added an equivalent raw property), This is the opposite to the approach @mtakac took in da82c96.

@ulferts @jpassing @NobodysNightmare if you have a preference, please let me know.

**Comment by ulferts** [comment]
@myabc I prefer having an html_[attribute_name] attribute so that clients might guess that they would have to render it as html and not as plaintext. 

The non html_ prefixed attribute will on the other hand be pretty useless as there is probably no client who knows how to render our "textile". @NobodysNightmare and @jpassing take on that would interest me here. They will probably have to render the html version as well because they will otherwise have unintelligible text. 

As long as we do not stick to a standard with the raw text the HTML version is the only standard we follow. A client might be able to strip the tags and come up with plaintext. We might also be able to but that would remove the styling.

So in the end we might remove the raw version output form the API for now and stick to alway returning HTML removing the necessity of having html_\* prefixed attributes at all. But we might want to keep them if we ever plan to have something intelligible without some magic by the client.

And @myabc just pointed out correctly that we will need the client to understand the "textile" if he is to correctly update a work package. However they will do that... So my final verdict is:
- html_[attribute_name] for the html version
- [attribute_name] for textile version

**Comment by jpassing** [comment]
I do not really care about the naming -- but it is important to offer both, html and textile.

**Comment by NobodysNightmare** [comment]
I think we will at least need attributes as raw (i.e. Textile) values, to be able to perform a roundtrip. Everything else is additional, e.g. HTML, plain etc.

I am not sure if my proposal conforms to the standards we are using for the API (e.g. HAL), but how about:

``` JSON
myAttribute: {
  raw: "This is *my* attribute",
  html: "This is <strong>my</strong> attribute",
  plain: "This is my attribute"
}
```

I think this would be more flexible than name-prefixes. Maybe we could even rename `raw` to `textile`.

Other topic regarding different attribute formats (and tell me if it is out-of-scope):
How does the client know that it will only be able to set `myAttribute` in its raw/textile form? And what happens if the client tries to do set e.g. `html_myAttribute`?

**Comment by myabc** [comment]
> I am not sure if my proposal conforms to the standards we are using for the API (e.g. HAL)

I don't think the standard says much about nested properties. I quite like this approach though (and it allows a degree of flexibility in the future).

**Comment by myabc** [comment]
@NobodysNightmare unfortunately, we had to go ahead and merge _as-is_. Your points won't be disregarded though: I'll create a new PR in the specification repo there, and we can continue discussion there.

**Pull request merged**